src/impl/format.ts

```diff
@@ -35,26 +36,38 @@ export function format(documentText: string, range: Range | undefined, options:
 		rangeEnd = documentText.length;
 	}
 	const eol = getEOL(options, documentText);
+	const eolFastPathSupported = supportedEols.includes(eol as any);
 
 	let numberLineBreaks = 0;
 
 	let indentLevel = 0;
 	let indentValue: string;
 	if (options.insertSpaces) {
-		indentValue = repeat(' ', options.tabSize || 4);
+		indentValue = cachedSpaces[options.tabSize || 4] ?? repeat(cachedSpaces[1], options.tabSize || 4);
```

Comment:
the `??` operator is a breaking change for node 12. since there's no engines declaration, everything that worked in 3.0.0 is part of the public API.

Reply:
The minimum version of this package is 14.x, so I think this is expected, it worked before because of luck not because was the intention.

An PR can be created to fix this but probably this can happen more times in the future.

Reply:
Yes, the minimum is now 16.x.

Reply:
Then this should have been a v4.0.0 that included an explicit engines.node declaration.